hash-parser/src/hash.pest

```diff
@@ -2,443 +2,453 @@
-///
-/// trait eq = <T> => (T, T) => bool;
-///     ┌─^^ ^─┐   ^─ ─ ─ ─ ─ ─ ─ ┐
-///   name   Generic type args    Function type definition
-trait_def   = {"trait" ~ ident ~ "=" ~ trait_bound ~ "=>" ~ type_t}
-
-bound       = {ident ~ type_args}
-type_args   = {"<" ~ type_t ~ (comma ~ type_t)* ~ ">"}
-trait_bound = {type_args ~ ("where" ~ (bound ~ (comma ~ bound)*))?}
-
-/// Intrinsic parsing, prefixed by '#' some identifier. There must be no spaces between
-/// the '#' and the ident, hence this grammar is atomic.
-intrinsic = @{"#" ~ ident}
-
-/// Parse a property access (period followed by a name)
-index           = {"[" ~ expr ~ "]"}
-property_access = {ident ~ index? ~ ("." ~ (ident ~ index?))*} // @@Cleanup: probably don't want to check for infix function calls here
-
-
-/// Grammar representation for a funtction call with the potential for specifying
-/// generic arguments and a list of actual arguments to the function. For example:
-///
-/// function_call<int, str>(some, arguments, for, func);
-///              ^^^^^^^^^^ ^---------^-------^-----^
-///           Type arguments to call        Function call arguments
-function_call = {callable_expr ~ type_args? ~ "(" ~ expr? ~ (comma ~ expr)* ~ ")"}
-
-/// Callable expr is used as a workaround for pest-LR limitation
-callable_expr = {
-    block                     |
-    intrinsic                 |
-    property_access           |
-    "(" ~ function_call ~ ")" | // @@Cleanup: find a way around having to add braces to expr
-    access_name               
+/// Expression which can be the subject of a function call/prop-access/index.
+subject_expr = {
+    import_expr   |
+    literal_expr  |
+    variable_expr |
+    paren_expr
 }
 
+/// Index
+index_arg = { square_left ~ expr ~ square_right }
 
-/// Loop constructs
-loop_statement  = {"loop" ~ block}
-while_statement = {"while" ~ expr ~ body_block}
-for_statement   = {"for" ~ exhaustive_pattern ~ "in" ~ expr ~ body_block}
+/// Function call
+fn_args = { paren_left ~ expr? ~ (comma_op ~ expr)* ~ paren_right }
 
-/// patterns sub grammars 
-struct_pattern    = {ident ~ "{" ~ (ident ~ (comma ~ ident)*)? ~ "}"}
-namespace_pattern = {"{" ~ (name ~ (comma ~ name)*)? ~ "}"}
-enum_pattern      = {ident ~ "(" ~ pattern ~ (comma ~ pattern)* ~ ")" }
+/// Property access
+property_access = { dot_op ~ name }
 
-/// Exhaustive patterns are those that are guaranteed to have a finite
-/// number of variants (level 1)
-exhaustive_pattern = {
-    ident             | // binding pattern
-    struct_pattern    |
-    namespace_pattern |
-    ("(" ~ (name ~ (comma ~ name)*)? ~ ")") // tuple pattern
-}
+/// Single term expression
+single_expr = { subject_expr ~ (property_access | fn_args | index_arg)* }
+
+/// Unary expression
+unary_expr = { unary_op ~ unary_expr | single_expr }
+
+/// Typed expression (as)
+typed_expr = { unary_expr ~ ( as_kw ~ any_type )? }
+
+/// Binary expression
+binary_expr = { typed_expr ~ ( binary_op ~ typed_expr )* }
```

Comment:
From a first glance this looks confusing, but essentially this creates precedence using pest's rules.

Reply:
Well this doesn't actually handle precedence, I figured we can use `PrecClimber` for that.

Reply:
But yes, it does handle precedence with `as` and other binary expressions.